src/hostname_resolution.py

```diff
@@ -60,111 +59,54 @@ def _update_host_details_in_databag(self, _) -> None:
             return
 
-        if not self._does_etc_hosts_need_update(host_details):
-            logger.debug("No hostnames in /etc/hosts changed. Skipping update to /etc/hosts")
-            return
-
-        hosts_in_file = []
-
-        with io.StringIO() as updated_hosts_file:
-            with open("/etc/hosts", "r") as hosts_file:
-                for line in hosts_file:
-                    if "# unit=" not in line:
-                        updated_hosts_file.write(line)
-                        continue
-
-                    for hostname, details in host_details.items():
-                        if hostname == line.split()[2]:
-                            hosts_in_file.append(hostname)
-
-                            fqdn, ip, unit = details["fqdn"], details["ip"], details["unit"]
-
-                            logger.debug(
-                                f"Overwriting {hostname} ({unit=}) with {ip=}, {fqdn=} in /etc/hosts"
-                            )
-                            updated_hosts_file.write(f"{ip} {fqdn} {hostname} # unit={unit}\n")
-                            break
-
-            for hostname, details in host_details.items():
-                if hostname not in hosts_in_file:
-                    fqdn, ip, unit = details["fqdn"], details["ip"], details["unit"]
-
-                    logger.debug(f"Adding {hostname} ({unit=} with {ip=}, {fqdn=} in /etc/hosts")
-                    updated_hosts_file.write(f"{ip} {fqdn} {hostname} # unit={unit}\n")
-
-            with open("/etc/hosts", "w") as hosts_file:
-                hosts_file.write(updated_hosts_file.getvalue())
-
-        try:
-            self.charm._mysql.flush_host_cache()
-        except MySQLFlushHostCacheError:
-            self.charm.unit.status = BlockedStatus("Unable to flush MySQL host cache")
-
-    def _remove_host_from_etc_hosts(self, event: RelationDepartedEvent) -> None:
-        departing_unit_name = event.unit.name
-
-        logger.debug(f"Checking if an entry for {departing_unit_name} is in /etc/hosts")
-        with open("/etc/hosts", "r") as hosts_file:
-            for line in hosts_file:
-                if f"# unit={departing_unit_name}" in line:
-                    break
-            else:
-                return
-
-        logger.debug(f"Removing entry for {departing_unit_name} from /etc/hosts")
-        with io.StringIO() as updated_hosts_file:
-            with open("/etc/hosts", "r") as hosts_file:
-                for line in hosts_file:
-                    if f"# unit={departing_unit_name}" not in line:
-                        updated_hosts_file.write(line)
-
-            with open("/etc/hosts", "w") as hosts_file:
-                hosts_file.write(updated_hosts_file.getvalue())
+        logger.debug("Updating /etc/hosts with new hostname to IP mappings")
+        hosts = Hosts()
+        hosts.remove_all_matching(comment=COMMENT)
+        # remove MAAS entry
+        hosts.remove_all_matching(address="127.0.1.1")
```

Comment:
This can bite more then just MAAS. Can we isolate impact to MAAS only?

Reply:
89a4e359